[FLINK-12159]. Enable YarnMiniCluster integration test under non-secure mode

zjffdu opened the pull request:
## What is the purpose of the change
I can not run flink in YarnMiniCluster for integration test because yarn-site.xml is not shipped to yarn container (krb5Config must be non-empty otherwise yarn-site.xml won't been shipped). It doesn't make sense to only allow yarn integration test under secure mode. So this PR just try to enable YarnMiniCluster integration test under non-secure mode. I tested this PR for zeppelin flink integration test.

## Brief change log

Straightforward code refactoring to separate `yarn-site.xml` and `krb5Config` file.


## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.
I verified this PR in zeppelin flink integration test.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no )
  - The S3 file system connector: (no )

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zjffdu commented:
@tillrohrmann  Could you help review it ? Thanks

zjffdu commented:
Thanks for the review @tillrohrmann . `yarn-site.xml` contains the RM scheduler address of `MiniCluster`.  If I don't ship `yarn-site.xml` to `MiniCluster`, JM can not connect to RM to allocate containers for TM.

tillrohrmann commented:
You mean that Flink's RM cannot talk to Yarn's RM without having the Yarn RM scheduler address, right? Does this mean that we don't have tests without Kerberos activated?

zjffdu commented:
@tillrohrmann I checked the `flink-yarn-tests` module, we have yarn test in non-secure mode. But the tricky thing is that we set `yarn.minicluster.fixed.ports` to `true` (https://github.com/apache/flink/blob/master/flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java#L169), so that JM can connect to Yarn RM without `yarn-site.xml`, because yarn RM will always use `8030` as the scheduler service port.
But I don't think this is the correct way to do integration test.
1. Using fixed port is easy to cause port conflicts in their local environment and it is hard for user to figure out why their test fail.
2. It is impossible to start 2 MiniYarnCluster.
3. It is hard for downstream project to follow this. We can not enforce the downstream project set `yarn.minicluster.fixed.ports` to `true` (The default value is `false`, so most of time user don't know he needs to set it to `true`)

I will update the test code as well.

tillrohrmann commented:
Thanks for looking into this @zjffdu. I agree with you that setting the port to a fixed value is not ideal for reason 1 and 2. Do you run into one of these problems?

What I don't fully understand is problem 3. If a downstream process sets up a `YarnMiniCluster` it could configure the port differently. Flink would not need to be changed. If using the `YarnTestBase`, then all other applications running on this cluster would need to use the default scheduler port.

Currently, your changes break the existing tests.

zjffdu commented:
> If a downstream process sets up a YarnMiniCluster it could configure the port differently. Flink would not need to be changed. 
If user use YarnMiniCluster directly, and don't set `yarn.minicluster.fixed.ports` to be `true`, then JM can not connect RM (JM will connect to default scheduler port 8030, but the actual scheduler port is a random port which is stored in yarn-site.xml and flink didn't ship it to JM).

@tillrohrmann I have fixed the UT failure issue. The root cause is that if I remove the setting of `yarn.minicluster.fixed.ports`, flink client can not connect RM to allocate container for JM because it didn't create yarnClient properly.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -449,7 +449,14 @@ public void addGeneralOptions(Options baseOptions) {
 		final Configuration effectiveConfiguration = applyCommandLineOptionsToConfiguration(commandLine);
-
+		// for unit tests only
+		if (System.getenv("IN_TESTS") != null) {
```
Instead of leaking the `IN_TESTS` hack out of the `AbstractYarnClusterDescriptor`, it would be better if the production and test code used the same code path. So for example, the `YarnTestBase` should setup the environment in such a way that the normal `FlinkYarnSessionCli` code finds the testing `YarnConfiguration`. That way, we would not have to insert special environment variables which signal that we are in a test.

An idea could be to modify the HADOOP_HOME env variable which is used when loading the Hadoop configuration. Maybe there is some other ENV variable which is respected by the YarnConfiguration which one could configure.

I think this would be a better solution to the problem. WDYT?

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -877,6 +867,18 @@ public ApplicationReport startAppMaster(
 		Path remoteYarnSiteXmlPath = null;
 		boolean hasKrb5 = false;
 		if (System.getenv("IN_TESTS") != null) {
+			File f = new File(System.getenv("YARN_CONF_DIR"), Utils.YARN_SITE_FILE_NAME);
```
Are we sure that this file always exists? What happens if this file does not exist? I think it would be best if we keep the test code as close to the production code as possible. In this case it would mean that we set the environment of the Yarn container up in a way that we don't have to explicitly ship the `yarn-site.xml` to it, if possible.

zjffdu commented:
@tillrohrmann  I agree with you that test and production should use the same code path. 
Yarn would try to find `yarn-site.xml` under classpath, so for current flink's unit test, we have to add `yarn-site.xml` explicitly.  

AFAIK, we have to do 2 things to make test code and production use the same code path:
1. Make sure flink client can connect to Yarn RM. For this, I can think of 2 approaches:
   a. Instead of running flink job via api, we could launch flink job in separated process via `flink run` and set `YARN_CONF_DIR` when using `flink run` script.
   b. After starting YarnMiniCluster, copy `yarn-site.xml` to `flink-yarn-tests/target/test-classes` so that flink client can load `yarn-site.xml`.

2. Make sure JM can connect to Yarn RM.  I am afraid this maybe a bug of `YarnMiniCluster` that its NM didn't put `yarn-site.xml` under its container's CLASSPATH (Or maybe this is not a bug of `YarnMiniCluster`, because when starting NM, there's no `yarn-site.xml`, it is generated after RM/NM is started, this is different from real yarn cluster). So the only approach for us is we always ship `yarn-site.xml` to JM.

zjffdu commented:
ping @tillrohrmann  Any thoughts ?

tillrohrmann commented:
I think you are right with your analysis:

1. The client needs to be able to connect to the started Yarn cluster.
2. The JM needs to be able to talk to Yarn's RM.

What we should not do is to launch a job via `flink run`. I think we should check whether this is truly a problem of Yarn. I would assume that other projects have the same problem and there should be a Yarn ticket for this problem. Maybe they also propose a solution for this problem.

Only if this is not the case I would start looking into copying our own `yarn-site.xml` to the required places.

zjffdu commented:
@tillrohrmann For point 2, I think either we keep `IN_TESTS` or we just always ship yarn-site.xml no matter it is test code or production code (This is what spark did). I don't think this is a bug of MiniYarnCluster, it is just a limitation of it. 
For point 1, I am afraid the only approach is just coping yarn-site.xml into 'target/test-classes'.

zjffdu commented:
@tillrohrmann  I have updated the PR. Now the only difference between test code and production code is in `AbstractYarnClusterDescriptor.java` where we check `IN_TESTS`.  I think we can remove it by always shipping `yarn-site.xml` just as what spark did. I don't think it is a large effect for job submission.  What do you think ?

tillrohrmann commented:
Ok, what about the following approach:

When we start the `YarnMiniCluster` we extract the `YarnConfiguration` via `YarnMiniCluster#getConfiguration` and write it as `yarn-site.xml` to `target/test-classes` or we do it like [here](https://github.com/uzh/signal-collect-yarn/blob/master/src/main/scala/com/signalcollect/deployment/yarn/MiniCluster.scala) where on has a dummy `yarn-site.xml` under resources which gets populated. 

The important bit is that this should get cleaned up at the end of the test. For this, it would be super handy if we had a `YarnResource` which extends `ExternalResource` where we have the `before` and `after` methods. We could think about adding a new component and bit by bit deprecating the `YarnTestBase`.

Next, we change the `YarnTestBase#createYarnClusterDescriptor` to add the `yarn-site.xml` file as a file to be shipped to the cluster. Then every test which deploys a Flink application should use this method to create the `YarnClusterDescriptor`.

tillrohrmann commented:
My proposal would be to explicitly add the `yarn-site.xml` to the set of files which are being shipped via the `YarnClusterDescriptor.addShipFiles` method when we call `YarnTestBase#createYarnClusterDescriptor`.

zjffdu commented:
@tillrohrmann I am afraid that just adding yarn-site.xml explicitly via addShipFiles is not sufficient. It will also be used by JM to create TaskManager. See the following link

- https://github.com/apache/flink/blob/master/flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java#L897
- https://github.com/apache/flink/blob/master/flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java#L970

And I found many unit test related code in flink-yarn/src/main which doesn't make sense to me. I think we do need to refactor the flink-yarn module. But I think this could be done in a follow-up ticket. Does it make sense to you ? For this PR, do you have any other concerns ? Thanks

https://github.com/apache/flink/blob/master/flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java#L78

tillrohrmann commented:
I think all shipped files will be added to the class path. Consequently, there should not be the need to set the `ENV_YARN_SITE_XML_PATH` env variable. This can be completely removed given that the `yarn-site.xml` can be found on the class path.

I think that we should not change the `flink-yarn` module in the context of this PR.

zjffdu commented:
@tillrohrmann  I could do other refactoring/improvement in another followup pr. Thanks for the review. Do you have any other suggestion for the current PR ?

tillrohrmann commented:
Have you verified that my comments are not working @zjffdu? If not, then I would still suggest to explicitly add the `yarn-site.xml` to the files which are being shipped to the cluster. That way we could get rid of the `IN_TESTS` env variable.

zjffdu commented:
@tillrohrmann  I have updated PR (add yarn-site.xml explicitly) , could you help review it ? Thanks

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -869,41 +859,40 @@ public ApplicationReport startAppMaster(
+			yarnSitePath,
+			localResources,
+			homeDir,
+			"");
```
This is not what I proposed. We should not add the `yarn-site.xml` per default to all Yarn deployments. Instead we should only add it if we are using the `YarnTestBase`. Therefore, I suggest to add 
```
yarnClusterDescriptor.addShipFiles(Arrays.asList(flinkLibFolder, yarnSiteXMLFile));
```

to `YarnTestBase#createYarnClusterDescriptor`.

zjffdu replied:
@tillrohrmann Unfortunately, not all yarn related test (e.g. YARNSessionFIFOITCase#testDetachedMode) use `YarnTestBase#createYarnClusterDescriptor`, so I can not only add yarn-site.xml explicitly in YarnTestBase

tillrohrmann replied:
I guess one could then refactor the `YarnTestBase#Runner` to instantiate the `FlinkYarnSessionCli` with a `Supplier<YarnClusterDescriptor>` which could setup the `YarnClusterDescriptor` accordingly.

This might be out of scope for this PR but maybe this test case should not test command line parsing and deploying a Yarn application. Instead one could this split up into two tests: One tests the command line parsing (`CliFrontend`) and the other test tests the detached mode.

zjffdu replied:
> I guess one could then refactor the YarnTestBase#Runner to instantiate the FlinkYarnSessionCli with a Supplier<YarnClusterDescriptor> or factory which could setup the YarnClusterDescriptor accordingly.

@tillrohrmann IMHO, this is not a good solution as well. Because it would also introduce different code path for test code and production code in `FlinkYarnSessionCli`. 
I totally agree that we need to refactor the flink client api, currently the command line parsing and cluster deployment is coupled together closely, which make it difficult to do unit testing (The test code in flink-yarn-tests module is not readable to me, which imply the issues of current flink api design).  I do have a plan to refactor the flink client api, but it would involve a large amount of work. 
For this PR, it is just to fix the YarnMiniCluster integration test bug so that downstream project can do integration test with flink. I think we can just fix the bug without touching the client api (keep `IN_TESTS`) for now. The refactoring of flink client api could be done in other PRs. Does it make sense to you ?

tillrohrmann replied:
Ok, I agree that this would widen the scope quite a bit. So the idea would then simply to always ship the yarn-site.xml file if `IN_TESTS` is set, right? Additionally, we will write the `yarn-site.xml` to the `test-classes` directory.

Please open the corresponding JIRA issues to properly fix this in the future.

zjffdu replied:
That's correct, I have created https://issues.apache.org/jira/browse/FLINK-12422 for the followup work.

tillrohrmann replied:
Thanks!

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -869,41 +859,40 @@ public ApplicationReport startAppMaster(
+				localResources,
+				homeDir,
+				"");
+			hasKrb5 = true;
```
Shipping the Kerberos files via explicitly calling `addShipFiles` would be good as well. Then we could get rid of this code here.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -966,8 +955,10 @@ public ApplicationReport startAppMaster(
+		}
+		if (remoteKrb5Path != null) {
 			appMasterEnv.put(YarnConfigKeys.ENV_KRB5_PATH, remoteKrb5Path.toString());
 		}
```
Would be good to get rid of these things as well if both files are on the classpath anyway.

tillrohrmann replied:
Might be out of scope though.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -432,17 +432,18 @@ static ContainerLaunchContext createTaskExecutorContext(
 		LocalResource krb5ConfResource = null;
 		boolean hasKrb5 = false;
-		if (remoteYarnConfPath != null && remoteKrb5Path != null) {
+		if (remoteYarnConfPath != null) {
```
This block should no longer be needed, hopefully.

zjffdu commented:
@tillrohrmann I have updated the PR, now I will ship the yarn-site.xml file if IN_TESTS is set, and write the yarn-site.xml to the test-classes directory. Let me know if you have any other concern

tillrohrmann commented:
Thanks a lot for the update @zjffdu. I will take a look.
